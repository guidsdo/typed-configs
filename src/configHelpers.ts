import * as fs from "fs";
import * as path from "path";
import * as Yaml from "yamljs";
import { ConfigValueType } from "./types";

const currentDir = path.resolve(".");

export function loadConfigfromYaml(configPath: string, isRequired = false): Record<string, any> {
    const cfgPathCurrentDir = path.join(currentDir, configPath);
    const checkedConfigPath = fs.existsSync(configPath) ? configPath : fs.existsSync(cfgPathCurrentDir) ? cfgPathCurrentDir : undefined;

    if (!checkedConfigPath) {
        if (isRequired) {
            throw new Error(`Configuration file '${configPath}' does not exist`);
        }

        return {};
    }

    return Yaml.load(configPath);
}

export function getEnvironmentVariableKeys(filterEmptyValues = false): string[] {
    return filterEmptyValues
        ? Object.entries(process.env)
              .filter(([, value]) => value != "")
              .map(([key]) => key)
        : Object.keys(process.env);
}

export function loadEnvironmentVariable(key: string, expectedType: ConfigValueType) {
    return checkValue(process.env, key, expectedType);
}

export function checkValue(envObject: NodeJS.ProcessEnv, key: string, expectedType: ConfigValueType) {
    const value = envObject[key];

    if (value === undefined) return undefined;

    // All env variables are set as strings, so we need to check first if we can parse them
    switch (expectedType) {
        case "boolean":
            if (["true", "false"].includes(value)) return value === "true";
            break;
        case "number":
            if (value !== "" && !Number.isNaN(Number(value))) return Number(value);
            break;
        case "string":
            return value;
    }

    throw new Error(`Invalid ENV value of property '${key}', which should be of type '${expectedType}' but got '${value}'.`);
}
