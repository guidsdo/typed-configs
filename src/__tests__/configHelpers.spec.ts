import { checkValue } from "../configHelpers";

describe("configHelpers", () => {
    describe("checkValue", () => {
        it("should return undefined when key isn't in the env object", () => {
            // expect(() => checkValue({}, "TEST_KEY", "string")).toThrow("Target ExampleClass doesn't have any config fields.");
            expect(checkValue({}, "TEST_KEY", "string")).toBeUndefined();
        });

        describe("when expecting a string", () => {
            it("should return a string when either 'true' or 'false' is passed", () => {
                expect(checkValue({ TEST_KEY: "true" }, "TEST_KEY", "string")).toBe("true");
                expect(checkValue({ TEST_KEY: "false" }, "TEST_KEY", "string")).toBe("false");
            });

            it("should return a string when a number is given", () => {
                expect(checkValue({ TEST_KEY: "8080" }, "TEST_KEY", "string")).toBe("8080");
                expect(checkValue({ TEST_KEY: "-1" }, "TEST_KEY", "string")).toBe("-1");
            });
        });

        describe("when expecting a boolean", () => {
            it("should return a boolean when either 'true' or 'false' is passed", () => {
                expect(checkValue({ TEST_KEY: "true" }, "TEST_KEY", "boolean")).toBe(true);
                expect(checkValue({ TEST_KEY: "false" }, "TEST_KEY", "boolean")).toBe(false);
            });

            it("should throw when a non-valid string boolean is given", () => {
                expect(() => checkValue({ TEST_KEY: "treu" }, "TEST_KEY", "boolean")).toThrowErrorMatchingSnapshot();
                expect(() => checkValue({ TEST_KEY: "0" }, "TEST_KEY", "boolean")).toThrowErrorMatchingSnapshot();
                expect(() => checkValue({ TEST_KEY: "1" }, "TEST_KEY", "boolean")).toThrowErrorMatchingSnapshot();
            });
        });

        describe("when expecting a number", () => {
            it("should return a number when a valid number string is passed", () => {
                expect(checkValue({ TEST_KEY: "8080" }, "TEST_KEY", "number")).toBe(8080);
                expect(checkValue({ TEST_KEY: "-1" }, "TEST_KEY", "number")).toBe(-1);
                expect(checkValue({ TEST_KEY: "1.1" }, "TEST_KEY", "number")).toBe(1.1);
            });

            it("should throw when a non-valid number string is given", () => {
                expect(() => checkValue({ TEST_KEY: "not-a-number" }, "TEST_KEY", "number")).toThrowErrorMatchingSnapshot();
                expect(() => checkValue({ TEST_KEY: "one" }, "TEST_KEY", "number")).toThrowErrorMatchingSnapshot();
                expect(() => checkValue({ TEST_KEY: "NaN" }, "TEST_KEY", "number")).toThrowErrorMatchingSnapshot();
                expect(() => checkValue({ TEST_KEY: "1 1" }, "TEST_KEY", "number")).toThrowErrorMatchingSnapshot();
            });
        });
    });
});
