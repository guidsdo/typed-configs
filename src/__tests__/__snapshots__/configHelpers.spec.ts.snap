// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configHelpers checkValue when expecting a boolean should throw when a non-valid string boolean is given 1`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'boolean' but got 'treu'."`;

exports[`configHelpers checkValue when expecting a boolean should throw when a non-valid string boolean is given 2`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'boolean' but got '0'."`;

exports[`configHelpers checkValue when expecting a boolean should throw when a non-valid string boolean is given 3`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'boolean' but got '1'."`;

exports[`configHelpers checkValue when expecting a number should throw when a non-valid number string is given 1`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'number' but got 'not-a-number'."`;

exports[`configHelpers checkValue when expecting a number should throw when a non-valid number string is given 2`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'number' but got 'one'."`;

exports[`configHelpers checkValue when expecting a number should throw when a non-valid number string is given 3`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'number' but got 'NaN'."`;

exports[`configHelpers checkValue when expecting a number should throw when a non-valid number string is given 4`] = `"Invalid ENV value of property 'TEST_KEY', which should be of type 'number' but got '1 1'."`;
